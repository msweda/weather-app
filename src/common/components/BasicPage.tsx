import {
  createStyles,
  Theme,
  WithStyles,
  withStyles,
} from '@material-ui/core/styles'
import * as React from 'react'

const BasicPage: React.FC<Props> = ({ classes, children }) => {
  return (
    <div className={classes.basicPage}>
      <div className={classes.leftPageMargin} />
      <div className={classes.pageContent}>{children}</div>
      <div className={classes.rightPageMargin} />
    </div>
  )
}

const styles = (theme: Theme) =>
  createStyles({
    basicPage: {
      flex: 1,
      display: 'flex',
      flexDirection: 'row',
      justifyContent: 'center',
      alignItems: 'stretch',
      backgroundColor: theme.palette.background.paper,
    },
    leftPageMargin: {
      flex: 1,
    },
    pageContent: {
      display: 'flex',
      flexDirection: 'column',
      justifyContent: 'center',
      alignItems: 'stretch',
      width: 800,
      [theme.breakpoints.down('sm')]: {
        width: 600,
      },
      [theme.breakpoints.down('xs')]: {
        width: 300,
      },
      padding: 20,
    },
    rightPageMargin: {
      flex: 1,
    },
  })

interface Props extends WithStyles<typeof styles> {}

export default withStyles(styles)(BasicPage)
