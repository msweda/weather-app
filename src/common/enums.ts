export enum RootRoute {
  Home = '/',
  Weather = '/weather',
  Forecast = '/forecast',
}

export enum AsyncStatus {
  Idle = 'IDLE',
  InProgress = 'IN_PROGRESS',
  Succeeded = 'SUCCEEDED',
  Failed = 'FAILED',
}
