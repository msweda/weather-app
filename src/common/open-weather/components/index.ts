export { default as UnitSelect } from './UnitSelect'
export { default as Weather } from './Weather'
