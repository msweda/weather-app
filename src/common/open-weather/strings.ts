export const noGeolocation = 'Geolocation must be enabled to use this feature.'
export const weatherFailed =
  'Weather information is not available at this time.'
export const forecastFailed =
  'Forecast information is not available at this time.'
export const weatherIconUrl = (weatherIcon: string) =>
  `http://openweathermap.org/img/w/${weatherIcon}.png`
