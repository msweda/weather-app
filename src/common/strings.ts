export const appHeading = 'Matthew Sweda\'s Weather App';
export const phoneNumber = '(248)-770-7487';
export const email = 'msweda.developer@gmail.com';
export const location = 'Detroit, MI';
export const copyright = (timestamp: number) => `Â© Matthew Sweda ${new Date(timestamp).getFullYear()}`;
export const emailLink = `mailto:${email}`;
export const linkedInLink = 'https://www.linkedin.com/in/matthew-sweda-118570ba/';
export const instagramLink = 'https://www.instagram.com/mattsweda/';