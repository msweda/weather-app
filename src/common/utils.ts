import createBrowserHistory from 'history/createBrowserHistory'

export const browserHistory = createBrowserHistory()

export function isGeolocationAvailable(): boolean {
  return 'geolocation' in navigator
}

export function getCurrentPosition() {
  return new Promise<Position>((resolve, reject) => {
    if (!isGeolocationAvailable()) {
      throw new Error('Geolocation is unavailable')
    }
    navigator.geolocation.getCurrentPosition(
      position => resolve(position),
      error => reject(error)
    )
  })
}
