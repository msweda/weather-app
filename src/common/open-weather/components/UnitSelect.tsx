import { UnitType } from '@/common/open-weather/enums'
import { AllStores } from '@/src/stores'
import FormControl from '@material-ui/core/FormControl'
import InputLabel from '@material-ui/core/InputLabel'
import MenuItem from '@material-ui/core/MenuItem'
import Select from '@material-ui/core/Select'
import {
  createStyles,
  Theme,
  WithStyles,
  withStyles,
} from '@material-ui/core/styles'
import { inject, observer } from 'mobx-react'
import * as React from 'react'
import OpenWeatherStore from '../store'

@inject((allStores: AllStores) => ({
  openWeatherStore: allStores.openWeatherStore,
}))
@observer
class UnitSelect extends React.Component<Props> {
  get injected() {
    return this.props as InjectedProps
  }

  render() {
    const { classes } = this.props
    return (
      <FormControl className={classes.selectFormControl}>
        <InputLabel htmlFor="weather-units-input">Units</InputLabel>
        <Select
          className={classes.select}
          value={this.injected.openWeatherStore.unitType}
          onChange={this.changeUnitType}
          inputProps={{ id: 'weather-units-input' }}
        >
          <MenuItem className={classes.menuItem} value={UnitType.Scientific}>
            Kelvin
          </MenuItem>
          <MenuItem className={classes.menuItem} value={UnitType.Metric}>
            Celsius
          </MenuItem>
          <MenuItem className={classes.menuItem} value={UnitType.Imperial}>
            Fahrenheit
          </MenuItem>
        </Select>
      </FormControl>
    )
  }

  private changeUnitType = (event: React.ChangeEvent<HTMLSelectElement>) => {
    this.injected.openWeatherStore.setUnitType(event.target.value as UnitType)
    this.injected.openWeatherStore.clearWeather()
    this.injected.openWeatherStore.clearForecast()
  }
}

const styles = (theme: Theme) =>
  createStyles({
    selectFormControl: {
      marginBottom: 20,
    },
    select: {
      color: theme.palette.primary.main,
    },
    menuItem: {
      color: theme.palette.primary.main,
    },
  })

interface Props extends WithStyles<typeof styles> {}

interface InjectedProps extends Props {
  openWeatherStore: OpenWeatherStore
}

export default withStyles(styles)(UnitSelect)
