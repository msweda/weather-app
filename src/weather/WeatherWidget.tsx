import { UnitSelect, Weather } from '@/common/open-weather/components'
import OpenWeatherStore from '@/common/open-weather/store'
import { noGeolocation, weatherFailed } from '@/common/open-weather/strings'
import { isGeolocationAvailable } from '@/common/utils'
import { AllStores } from '@/src/stores'
import Button from '@material-ui/core/Button'
import CircularProgress from '@material-ui/core/CircularProgress'
import { createStyles, WithStyles, withStyles } from '@material-ui/core/styles'
import Typography from '@material-ui/core/Typography'
import { inject, observer } from 'mobx-react'
import * as React from 'react'

@inject((allStores: AllStores) => ({
  openWeatherStore: allStores.openWeatherStore,
}))
@observer
class WeatherWidget extends React.Component<Props> {
  get injected() {
    return this.props as InjectedProps
  }

  render() {
    if (!isGeolocationAvailable()) {
      return <Typography>{noGeolocation}</Typography>
    }

    const { classes } = this.props

    return (
      <React.Fragment>
        <div className={classes.userInput}>
          <UnitSelect />
          {this.renderGetWeatherButton()}
        </div>
        {this.renderWeatherBlock()}
      </React.Fragment>
    )
  }

  private renderGetWeatherButton = () => {
    if (this.injected.openWeatherStore.weatherInProgress) {
      return <CircularProgress />
    }
    return (
      <Button variant="contained" color="primary" onClick={this.getWeather}>
        Get Weather
      </Button>
    )
  }

  private renderWeatherBlock = () => {
    if (
      this.injected.openWeatherStore.weatherSucceeded &&
      this.injected.openWeatherStore.weatherData !== undefined
    ) {
      const { classes } = this.props
      return (
        <div className={classes.weatherBlock}>
          <Typography variant="h6" color="textSecondary">
            The current temperature in your area is:
          </Typography>
          <Typography variant="h1" color="primary">
            {this.injected.openWeatherStore.weatherData.main.temp}
          </Typography>
          <Typography variant="h6" color="textSecondary">
            Conditions:
          </Typography>
          <Weather
            weather={this.injected.openWeatherStore.weatherData.weather}
          />
        </div>
      )
    }

    if (this.injected.openWeatherStore.weatherFailed) {
      return <Typography>{weatherFailed}</Typography>
    }

    return null
  }

  private getWeather = () => {
    this.injected.openWeatherStore.getWeather()
  }
}

const styles = () =>
  createStyles({
    userInput: {
      display: 'flex',
      flexDirection: 'row',
      justifyContent: 'center',
      alignItems: 'center',
      marginBottom: 20,
      '& > *': {
        margin: 10,
      },
    },
    weatherBlock: {
      textAlign: 'center',
    },
  })

interface Props extends WithStyles<typeof styles> {}

interface InjectedProps extends Props {
  openWeatherStore: OpenWeatherStore
}

export default withStyles(styles)(WeatherWidget)
