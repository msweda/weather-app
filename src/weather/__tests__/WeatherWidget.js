import { UnitSelect, Weather } from '@/common/open-weather/components'
import OpenWeatherStore from '@/common/open-weather/store'
import { noGeolocation, weatherFailed } from '@/common/open-weather/strings'
import { isGeolocationAvailable } from '@/common/utils'
import Button from '@material-ui/core/Button'
import CircularProgress from '@material-ui/core/CircularProgress'
import Typography from '@material-ui/core/Typography'
import { shallow } from 'enzyme'
import * as React from 'react'
import WeatherWidget from '../WeatherWidget'

jest.mock('@/common/open-weather/store')
jest.mock('@/common/utils')

describe('<WeatherWidget />', () => {
  let openWeatherStore
  let wrapper

  function createWrapper() {
    return shallow(
      <WeatherWidget.wrappedComponent
        openWeatherStore={openWeatherStore}
        classes={{}}
      />
    )
  }

  beforeEach(() => {
    openWeatherStore = new OpenWeatherStore()
    wrapper = createWrapper()
  })

  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('displays a disclaimer when geolocation is not available', () => {
    isGeolocationAvailable.mockImplementationOnce(() => false)
    wrapper = createWrapper()
    const typography = wrapper.find(Typography)
    expect(typography.exists()).toBeTruthy()
    expect(typography.children().text()).toEqual(noGeolocation)
  })

  describe('navigator.geolocation is defined', () => {
    beforeEach(() => {
      isGeolocationAvailable.mockImplementationOnce(() => true)
    })

    it('renders a <UnitSelect />', () => {
      wrapper = createWrapper()
      const unitSelect = wrapper.find(UnitSelect)
      expect(unitSelect.exists()).toBeTruthy()
    })

    describe('renderGetWeatherButton', () => {
      it('renders a <CircularProgress /> when weather is in progress', () => {
        openWeatherStore.weatherInProgress = true
        wrapper = createWrapper()
        const circularProgress = wrapper.find(CircularProgress)
        expect(circularProgress.exists()).toBeTruthy()
      })

      it('renders a get weather <Button /> when weather is not in progress', () => {
        openWeatherStore.weatherInProgress = false
        wrapper = createWrapper()
        const button = wrapper.find(Button)
        expect(button.exists()).toBeTruthy()
        expect(button.children().text()).toEqual('Get Weather')
      })
    })

    describe('renderWeatherBlock', () => {
      it('renders <Typography /> for temperature and <Weather /> when weather succeeded and is defined', () => {
        openWeatherStore.weatherInProgress = false
        openWeatherStore.weatherSucceeded = true
        openWeatherStore.weatherData = mockWeatherData
        wrapper = createWrapper()
        const typography = wrapper.find(Typography)
        expect(typography).toHaveLength(3)
        expect(
          typography
            .at(0)
            .children()
            .text()
        ).toEqual('The current temperature in your area is:')
        expect(
          typography
            .at(2)
            .children()
            .text()
        ).toEqual('Conditions:')
        const weather = wrapper.find(Weather)
        expect(weather.exists()).toBeTruthy()
        expect(weather.props().weather).toEqual(mockWeatherData.weather)
      })

      it('renders a disclaimer when weather fails', () => {
        openWeatherStore.weatherInProgress = false
        openWeatherStore.weatherFailed = true
        wrapper = createWrapper()
        const typography = wrapper.find(Typography)
        expect(typography.exists()).toBeTruthy()
        expect(typography.children().text()).toEqual(weatherFailed)
      })

      it('renders null when weather neither succeeded nor failed', () => {
        expect(wrapper.instance().renderWeatherBlock()).toBeNull()
      })
    })
  })
})

const mockWeatherData = {
  coord: {
    lon: -83.4,
    lat: 42.33,
  },
  weather: [
    {
      id: 602,
      main: 'Snow',
      description: 'heavy snow',
      icon: '13n',
    },
    {
      id: 721,
      main: 'Haze',
      description: 'haze',
      icon: '50n',
    },
    {
      id: 701,
      main: 'Mist',
      description: 'mist',
      icon: '50n',
    },
  ],
  base: 'stations',
  main: {
    temp: 37.38,
    pressure: 1017,
    humidity: 75,
    temp_min: 35.6,
    temp_max: 41.36,
  },
  visibility: 9656,
  wind: {
    speed: 12.75,
    deg: 210,
  },
  snow: {
    '1h': 76,
  },
  clouds: {
    all: 75,
  },
  dt: 1549151760,
  sys: {
    type: 1,
    id: 3869,
    message: 0.0047,
    country: 'US',
    sunrise: 1549197911,
    sunset: 1549234209,
  },
  id: 5014681,
  name: 'Westland',
  cod: 200,
}
