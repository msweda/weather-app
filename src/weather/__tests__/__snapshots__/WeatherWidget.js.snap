// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<WeatherWidget /> should match snapshot 1`] = `
<WithStyles(Typography)>
  Geolocation must be enabled to use this feature.
</WithStyles(Typography)>
`;
